src/ORM/DataList.php

```diff
@@ -658,6 +659,39 @@ public function filterByCallback(callable $callback): SS_List
         return $output;
     }
 
+    /**
+     * Return a copy of this list which only includes items where $fieldToFilterBy matches values in $fieldInOtherList from $list.
+     *
+     * If both fields are ID, the $list dataclass must be in the same class hierarchy as the dataclass in this list
+     *
+     * @param SS_List<DataObject> $list
+     * @param string $fieldToFilterBy The name of the field in $this to filter by
+     * @param string $fieldInOtherList The name of the field in $list to get filter values from
+     * @return static<T>
+     * @throws InvalidArgumentException
+     */
+    public function filterByList(SS_List $list, string $fieldToFilterBy = 'ID', string $fieldInOtherList = 'ID'): static
+    {
+        $thisDataClass = $this->dataClass();
+        $listDataClass = $this->getDataClassFromList($list);
+        if ($fieldInOtherList === 'ID'
+            && $fieldInOtherList === 'ID'
+            && !is_a($thisDataClass, $listDataClass, true)
+            && !is_a($listDataClass, $thisDataClass, true)
+        ) {
+            throw new InvalidArgumentException('If both columns are ID, the $list dataclass must be in the same class hierarchy as the dataclass in this list');
+        }
```

Comment:
This and the same exception in `excludeByList()` comes from the exception thrown by `subtract()`.
Since IDs don't directly match across different class hierarchies, it doesn't make sense to allow filtering/excluding by other class hierarchies since that probably represents a mistake rather than an intentional filter.

Note that I've widened this from `subtract()`'s check to allow classes anywhere in the same model hierarchy, not just the exact same data class. This means you can filter by a list of subclass records, for example.